File: backend/routes/location.py
"""Location utilities for browser-driven Explore features."""

from typing import Any

import httpx
from fastapi import APIRouter, HTTPException, Query, status

from backend.config import GOOGLE_API_KEY

router = APIRouter()

GOOGLE_GEOCODE_URL = "https://maps.googleapis.com/maps/api/geocode/json"
GEOCODE_TIMEOUT_SECONDS = 8.0


def _component_name(components: list[dict[str, Any]], component_type: str) -> str:
    for component in components:
        if component_type in (component.get("types") or []):
            return str(component.get("long_name") or component.get("short_name") or "").strip()
    return ""


def _location_label_from_result(result: dict[str, Any]) -> tuple[str, str, str]:
    components = result.get("address_components") or []
    if not isinstance(components, list):
        return "", "", ""

    city = (
        _component_name(components, "locality")
        or _component_name(components, "postal_town")
        or _component_name(components, "administrative_area_level_2")
    )
    region = _component_name(components, "administrative_area_level_1")

    if city and region:
        return city, region, f"{city}, {region}"
    if city:
        return city, "", city
    if region:
        return "", region, region
    return "", "", ""


@router.get("/location/reverse")
async def reverse_geocode_location(
    lat: float = Query(..., ge=-90, le=90),
    lng: float = Query(..., ge=-180, le=180),
):
    """Resolve coordinates to a city and province/state label."""
    if not GOOGLE_API_KEY:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Location lookup is not configured.",
        )

    params = {
        "latlng": f"{lat},{lng}",
        "result_type": "locality|administrative_area_level_1|postal_town",
        "key": GOOGLE_API_KEY,
    }

    try:
        async with httpx.AsyncClient(timeout=GEOCODE_TIMEOUT_SECONDS) as client:
            response = await client.get(GOOGLE_GEOCODE_URL, params=params)
            response.raise_for_status()
            payload = response.json()
    except httpx.HTTPError as exc:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="Location lookup failed.",
        ) from exc

    if payload.get("status") not in {"OK", "ZERO_RESULTS"}:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="Location lookup failed.",
        )

    for result in payload.get("results") or []:
        city, region, label = _location_label_from_result(result)
        if label:
            return {"city": city, "region": region, "label": label}

    raise HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail="Location could not be resolved.",
    )

File: backend/src/routes/location.rs
use crate::{
    errors::{AppError, Result},
    security,
    state::AppState,
};
use axum::{
    extract::{Query, State},
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use serde::Deserialize;
use serde_json::{json, Value};
use std::sync::Arc;

pub fn router() -> Router<Arc<AppState>> {
    Router::new().route("/location/reverse", get(reverse_geocode))
}

#[derive(Deserialize)]
struct ReverseGeoParams {
    lat: f64,
    lng: f64,
}

async fn reverse_geocode(
    State(state): State<Arc<AppState>>,
    Query(params): Query<ReverseGeoParams>,
) -> Result<impl IntoResponse> {
    security::validate_lat_lng(params.lat, params.lng)?;

    if state.config.google_api_key.is_empty() {
        let label = format!("{:.4}, {:.4}", params.lat, params.lng);
        return Ok(Json(json!({
            "formatted_address": label,
            "city": null,
            "state": null,
            "region": null,
            "label": label,
        })));
    }

    let url = format!(
        "https://maps.googleapis.com/maps/api/geocode/json?latlng={},{}&key={}",
        params.lat, params.lng, state.config.google_api_key
    );

    let resp: Value = reqwest::get(&url)
        .await
        .map_err(|e| AppError::Internal(e.to_string()))?
        .json()
        .await
        .map_err(|e| AppError::Internal(e.to_string()))?;

    let results = resp["results"].as_array().cloned().unwrap_or_default();
    if results.is_empty() {
        return Ok(Json(
            json!({ "formatted_address": null, "city": null, "state": null, "region": null, "label": format!("{:.4}, {:.4}", params.lat, params.lng) }),
        ));
    }

    let result = &results[0];
    let formatted = result["formatted_address"]
        .as_str()
        .unwrap_or("")
        .to_string();

    let components = result["address_components"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    let city = find_component(&components, "locality");
    let state_name = find_component(&components, "administrative_area_level_1");
    let label = match (&city, &state_name) {
        (Some(city), Some(region)) => format!("{}, {}", city, region),
        (Some(city), None) => city.clone(),
        (None, Some(region)) => region.clone(),
        (None, None) => formatted.clone(),
    };

    Ok(Json(json!({
        "formatted_address": formatted,
        "city": city,
        "state": state_name,
        "region": state_name,
        "label": label,
        "lat": params.lat,
        "lng": params.lng,
    })))
}

fn find_component(components: &[Value], component_type: &str) -> Option<String> {
    components.iter().find_map(|c| {
        let types = c["types"].as_array()?;
        if types.iter().any(|t| t.as_str() == Some(component_type)) {
            c["long_name"].as_str().map(String::from)
        } else {
            None
        }
    })
}
